src/gt4py/next/iterator/transforms/trace_shifts.py

```diff
@@ -26,6 +26,30 @@ class Sentinel(enum.Enum):
     TYPE = object()
 
 
+@dataclasses.dataclass(frozen=True)
+class ShiftRecorder:
+    recorded_shifts: dict[int, set[tuple[ir.OffsetLiteral, ...]]] = dataclasses.field(
+        default_factory=dict
+    )
+
+    def register_node(self, inp: ir.Expr | ir.Sym) -> None:
+        self.recorded_shifts.setdefault(id(inp), set())
+
+    def __call__(self, inp: ir.Expr | ir.Sym, offsets: tuple[ir.OffsetLiteral, ...]) -> None:
+        self.recorded_shifts[id(inp)].add(offsets)
```

Comment:
```suggestion
    def __call__(self, inp: ir.Expr | ir.Sym, offsets: tuple[ir.OffsetLiteral, ...]) -> None:
        self.recorded_shifts.setdefault(id(inp), set()).add(offsets)
```

Reply:
I thought the register mechanism is not nice, but it slightly changes the result: not in the result means no offset applied.

Reply:
As discussed TODO test case